dds/DCPS/DiscoveryBase.h

```diff
@@ -272,15 +461,15 @@ namespace OpenDDS {
           if (iter != topic_names_.end()) {
             ignored_topics_.insert(iter->second);
             // Remove all publications and subscriptions on this topic
-            typename OPENDDS_MAP(OPENDDS_STRING, TopicDetails)::iterator top_it =
-              topics_.find(iter->second);
-            if (top_it != topics_.end()) {
-              TopicDetails& td = top_it->second;
-              RepoIdSet::iterator ep;
-              for (ep = td.endpoints_.begin(); ep!= td.endpoints_.end(); ++ep) {
-                match_endpoints(*ep, td, true /*remove*/);
-                if (shutting_down()) { return; }
-              }
+            TopicDetails& td = topics_[iter->second];
+            RepoIdSet ids = td.endpoints();
+            for (RepoIdSet::iterator ep = ids.begin(); ep!= ids.end(); ++ep) {
+              match_endpoints(*ep, td, true /*remove*/);
+              td.remove_pub_sub(*ep);
+              if (shutting_down()) { return; }
```

Comment:
What prompts these checks for shutdown? Is something releasing & reacquiring a lock in here? Can the state change every time we call match_endpoints? If that's the case, maybe we need to be checking for shutdown immediately after match_endpoints, including the two calls above? Otherwise, why are we doing it here at all?

Reply:
That was pre-existing code.  We should follow up with @mitza-oci

Reply:
`match_endpoints(a, b, true)` calls `remove_assoc` which calls back to DCPS.  Please take a look at that code and let me know if there are specific questions I can answer.